src/microbe_stage/organelle_unlocks/StatisticBasedUnlockCondition.cs

```diff
@@ -0,0 +1,188 @@
+﻿namespace UnlockConstraints
+{
+    using Newtonsoft.Json;
+
+    /// <summary>
+    ///   An unlock condition that relies on a certain <see cref="IStatistic"/> to track progress
+    /// </summary>
+    public abstract class StatisticBasedUnlockCondition : IUnlockCondition
+    {
+        [JsonIgnore]
+        public abstract StatsTrackerEvent RelevantEvent { get; }
+
+        [JsonIgnore]
+        public IStatistic? RelevantStatistic { get; set; }
+
+        public abstract bool Satisfied();
+
+        public abstract void GenerateTooltip(LocalizedStringBuilder builder);
+
+        public virtual void Check(string name)
+        {
+        }
+
+        public virtual void Resolve(SimulationParameters parameters)
+        {
+        }
+    }
+
+    /// <summary>
+    ///   The number of microbes engulfed by the player is above the specified amount.
+    /// </summary>
+    public class EngulfedMicrobesAbove : StatisticBasedUnlockCondition
+    {
+        public override StatsTrackerEvent RelevantEvent => StatsTrackerEvent.PlayerEngulfedOther;
+
+        [JsonProperty]
+        public int Required { get; set; }
+
+        [JsonIgnore]
+        public SimpleStatistic? Engulfed => (SimpleStatistic?)RelevantStatistic;
+
+        public override bool Satisfied()
+        {
+            if (Engulfed == null)
+                return false;
+
+            return Engulfed.Value >= Required;
+        }
+
+        public override void GenerateTooltip(LocalizedStringBuilder builder)
+        {
+            if (Engulfed == null)
+                return;
+
+            builder.Append(new LocalizedString("ENGULFED_MICROBES_ABOVE", Required, Engulfed.Value));
+        }
+    }
+
+    /// <summary>
+    ///   The number of times the player has died is above a specified amount.
+    /// </summary>
+    public class PlayerDeathsAbove : StatisticBasedUnlockCondition
+    {
+        public override StatsTrackerEvent RelevantEvent => StatsTrackerEvent.PlayerDied;
+
+        [JsonProperty]
+        public int Required { get; set; }
+
+        [JsonIgnore]
+        public SimpleStatistic? Deaths => (SimpleStatistic?)RelevantStatistic;
+
+        public override bool Satisfied()
+        {
+            if (Deaths == null)
+                return false;
+
+            return Deaths.Value >= Required;
+        }
+
+        public override void GenerateTooltip(LocalizedStringBuilder builder)
+        {
+            if (Deaths == null)
+                return;
+
+            builder.Append(new LocalizedString("PLAYER_DEATH_COUNT_ABOVE", Required, Deaths.Value));
+        }
+    }
+
+    public class ReproduceInBiome : StatisticBasedUnlockCondition
+    {
+        [JsonIgnore]
+        private Biome biome = null!;
+
+        public override StatsTrackerEvent RelevantEvent => StatsTrackerEvent.PlayerReproduced;
+
+        [JsonProperty("Biome")]
+        public string RawBiome { get; set; } = string.Empty;
```

Comment:
Biome type as a registry type should be JSON loadable directly without a string here.

Reply:
You mean `BiomeType`? Well, the `Biome` class doesn't actually contain a `BiomeType` field.

Instead, I won't actually get the biome, just use the biome type.

Reply:
I'm a bit confused. Correct me if I'm wrong, but I don't think we can use the actual `Biome` type here. I tried and the json converter complained that it can't convert a string to a type of `Biome`. We're not trying to define new biomes here after all, just reference the already defined ones in `biomes.json` right?

In that case I believe the best option is to either use a string or use `BiomeType` as that contains all the data this unlock condition needs.

Reply:
That's one of those save tricky parts I mentioned. It should just work by default, like how organelles can be defined. Though there is a potential order of loading problem if the unlock conditions are retrieved from JSON, so the init order might need swapping in simulation parameters. Like I offered before, if you do the conversion to the "right" code I can take a look at getting the JSON loading wrangled into behaving nicely.

For dictionaries things get much more fun but a basic object reference should just work. As I already wrote this I'll share this bit of info with you: in case of dictionaries there's about 2 possible solutions, slapping `[TypeConverter(typeof(ThriveTypeConverter))]` on `Biome` or slapping `[JsonConverter(typeof(DictionaryWithJSONKeysConverter<Biome, float>))]` on the dictionary property. But again both of those should be unnecessary, it's much more likely to be an initialization order problem where the biome data is not yet available. That should be fixable as I don't foresee biome data referring back to organelle types which would cause a circular dependency.

Reply:
Good to know and I'm sure it might prove useful, though it is not what I was originally asking about. I wasn't referencing the dictionary in `ReproductionStatistic`, but that `Biome` cannot be loaded from the registry (not save files, this data is not even included in saves) as just a string. Replacing the field `RawBiome` with a field of type `Biome` throws an error that a sting cannot be converted into an object of type biome on registry load.

Reply:
I addressed that first in my comment. I think it is a data dependency ordering problem. If you can convert the code to what it should be here, I can look into the JSON deserialize problem.